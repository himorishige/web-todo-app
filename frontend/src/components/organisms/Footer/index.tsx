/** @jsxImportSource @emotion/react */
import { css } from '@emotion/react';
import { Button, Input } from 'src/components/atoms';
import { useToast } from 'src/hooks/useToast';
import { useForm, SubmitHandler } from 'react-hook-form';
import { useAppDispatch, useAppSelector } from 'src/app/hooks';
import { createTask, selectStatus } from 'src/features/tasks/tasksSlice';
import { useHistory, useLocation } from 'react-router-dom';

type Props = {};

type Inputs = {
  taskName: string;
};

const Footer: React.VFC<Props> = () => {
  const dispatch = useAppDispatch();
  const status = useAppSelector(selectStatus);
  const { showToast } = useToast();
  const history = useHistory();
  const location = useLocation();

  const {
    register,
    handleSubmit,
    watch,
    reset,
    formState: { errors },
  } = useForm<Inputs>();

  const onSubmit: SubmitHandler<Inputs> = async (data) => {
    // タスクの登録
    const result = await dispatch(
      createTask({
        title: data.taskName,
        description: '',
        priority: 0,
        isCompleted: false,
      }),
    );

    if (createTask.fulfilled.match(result)) {
      showToast('SUCCESS', 'タスクを登録しました');
      reset();
      if (location.pathname !== '/') {
        history.push('/');
      }
    }
    if (createTask.rejected.match(result)) {
      const message = result.error.message;
      showToast('FAIL', `タスクの登録に失敗しました ${message}`);
    }
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <div css={footerStyle}>
        <div css={inputWrapper}>
          <Input
            label="taskName"
            placeholder="タスクを入力してください"
            register={register}
            required
            disabled={status === 'loading'}
          />
        </div>
        <div css={buttonWrapper}>
          <Button primary label="登録" disabled={!watch('taskName') || status === 'loading'} />
        </div>
      </div>
    </form>
  );
};

export default Footer;

const footerStyle = css`
  background: var(--tertiary-color);
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 2rem 1.25rem;
`;

const inputWrapper = css`
  width: 100%;
`;

const buttonWrapper = css`
  padding-left: 1rem;
`;
