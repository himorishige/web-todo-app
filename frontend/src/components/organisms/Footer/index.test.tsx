import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { store } from 'src/app/store';
import Footer from '.';

describe('Organisms/Footer', () => {
  test('正しくレンダリングされている', () => {
    const element = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <Footer />
          </MemoryRouter>
        </Provider>,
      )
      .toJSON();
    expect(element).toMatchSnapshot();
  });
});
