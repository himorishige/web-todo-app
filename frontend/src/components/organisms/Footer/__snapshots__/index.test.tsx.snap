// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Organisms/Footer 正しくレンダリングされている 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="css-1qqqf83"
  >
    <div
      className="css-7d0ne1"
    >
      <div
        className="css-1ago99h"
      >
        <input
          className="css-5x3p9d"
          data-testid="input-area"
          disabled={false}
          maxLength={40}
          name="taskName"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="新しいタスクを入力してください"
          type="text"
        />
      </div>
      <div
        className="css-75wbz6"
      >
        <button
          className="css-k3l7h4"
          disabled={true}
        >
          登録
        </button>
      </div>
    </div>
  </div>
</form>
`;
